File: 713. Subarray Product Less Than K/NumSubarrayProductLessThanK.py
class Solution:
    def numSubarrayProductLessThanK(self, nums, k: int):
        n = len(nums)
        ans = 0
        
        l, r = 0, 0
        currProd = nums[0]
        while r < n:
            while currProd >= k:
                currProd /= nums[l]
                l +=1
            
            ans += r - l + 1
            r +=1
            if r < n:   currProd *= nums[r]

            
        return ans


sol = Solution()
print(sol.numSubarrayProductLessThanK([10,5,2,6], k = 100))

File: 713. Subarray Product Less Than K/num_subarray_product_less_than_k.rs
impl Solution {
    // O(n) time,
    // O(1) space,
    // Approach: sliding window
    pub fn num_subarray_product_less_than_k(nums: Vec<i32>, k: i32) -> i32 {
        if k <= 1 {
            return 0;
        }

        let n = nums.len();
        let (mut left, mut right) = (0, 0);
        let mut prod = 1;
        let mut ans = 0;

        while right < n {
            prod *= nums[right];

            while prod >= k {
                prod /= nums[left];
                left += 1;
            }
            right += 1;
            ans += (right - left) as i32;
        }

        ans
    }
}
